components/clp-mcp-server/clp_mcp_server/server/session_manager.py

```diff
@@ -0,0 +1,232 @@
+"""Session management for CLP MCP Server."""
+
+import threading
+import time
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta, timezone
+from typing import Any, ClassVar
+
+from paginate import Page
+
+from . import constants
+
+
+class PaginatedQueryResult:
+    """Paginates the cached log entries returned from a query's response."""
+
+    def __init__(self, result_log_entries: list[str], num_items_per_page: int) -> None:
+        """
+        Initializes the PaginatedQueryResult.
+
+        :param result_log_entries: List of cached log entries to paginate.
+        :param num_items_per_page:
+        :raise: ValueError if the number of cached results or num_items_per_page is invalid.
+        """
+        if len(result_log_entries) > constants.MAX_CACHED_RESULTS:
+            err_msg = (
+                f"PaginatedQueryResult exceeds maximum allowed cached results:"
+                f" {len(result_log_entries)} > {constants.MAX_CACHED_RESULTS}."
+            )
+            raise ValueError(err_msg)
+
+        if num_items_per_page <= 0:
+            err_msg = (
+                f"Invalid num_items_per_page: {num_items_per_page}, "
+                "it must be a positive integer. "
+            )
+            raise ValueError(err_msg)
+
+        self.result_log_entries = result_log_entries
+        self.num_items_per_page = num_items_per_page
+
+        self._num_pages = (len(result_log_entries) + num_items_per_page - 1) // num_items_per_page
+
+    def get_page(self, page_index: int) -> Page | None:
+        """
+        Returns a page from the cached query results.
+
+        :param page_index: The number of page to retrieve (zero-based index; 0 is the first page).
+        :return: A `Page` object for the specified page.
+        :return: None if `page_index` is out of bounds.
+        """
+        page_number = page_index + 1  # Convert zero-based to one-based
```

Comment:
```suggestion
        # Convert zero-based to one-based
        page_number = page_index + 1
```
We prefer to put the inline comment in a separate line.